### PR TITLE
384: Export the contracts

### PR DESCRIPTION
**Pull request by rgarner**
# Export the contracts

Add a `rake export:contracts` task. Map the fields from the JSONB field `SubmissionEntry#data` during export on a per-framework and per-`entry_type` basis. Perform no transformations on this data; there is an implicit assumption presently that values in the ingested `data` field are in their final form, though see "Data Issues".

## Data Issues

1. `Tier Number`, which feeds the export's `Lot Number` field is not present in 4 rows of contracts from a 20th Sep DB cut
1. Dates are in a variety of forms. As stated, no transformation is done on these
1. It may be the case that the Rail framework differs slightly from other two legal frameworks in the way it lays out its "additional" fields. **Mappings WILL require further work**.

### Commits 1-5

These tidy up the testing in general, making the top-level rake specs in `spec/lib/tasks` the effective features. The rake interface is also tidied up and de-duped. It's now possible to say `rake export:all` to create all the files in `/tmp`, or `rake export:invoices[stdout]` in order to pipe to `grep` for data quality checking purposes.

- moved integration tests for `Export::Invoices`, `Export::Tasks` and `Export::Submissions` to their respective rake task specs in `spec/lib/tasks`
- created an `Export::Anything` class that can output relations to an auto-generated file in `tmp` like `/tmp/tasks_2018-05-11.csv` or `:stdout`
- changed the rake tasks to use `Export::Anything` instead of doing everything in the rake task itself.

### Commits 6-

These are the meat of the export, and acknowledge that on Friday the [spreadsheet](https://docs.google.com/spreadsheets/d/1VD66SuFa-Q0MCE80j6PDBNuxgp7ElHX0lo-3UX6Bots/edit?ts=5ba274f7#gid=0) from which we have been working had a filter removed, revealing that mappings were not the same across orders and invoices.

There is now an `Export::Template` module, included by the `SubmissionEntryRow` derivatives `Export::Invoices::Row` and `Export::Contracts::Row` which gets the values for each framework/`SubmissionEntry#entry_type` via `value_for('TargetExportField')`. At present the mappings are stored in hash constants, but they will need to move out of here for October frameworks.

### CODE REVIEW

File: app/models/export/contracts/row.rb [hunk 1]
```diff
@@ -4,10 +4,14 @@ class Row < Export::CsvRow
       alias_method :contract, :model
 
       def row_values
-        puts contract.data.inspect
```
Comment (tekin) [inline app/models/export/contracts/row.rb]: ... yep, there it goes!

File: app/models/export/anything.rb [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    attr_reader :output_name
+
+    def initialize(relation, output_name = :tmp)
+      @relation = relation
+      @output_name = (output_name || :tmp).to_sym
+    end
+
+    def run
+      if @relation.empty?
+        STDERR.puts 'No records to export'
```
Comment (tekin) [inline app/models/export/anything.rb]: given that it's likely this code will be run with using the "all" rake task perhaps it would be useful for the user to know exactly which model there are no records for?

File: app/models/export/contracts/extract.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
+module Export
+  class Contracts
+    module Extract
```
Comment (tekin) [inline app/models/export/contracts/extract.rb]: looks like there are no spec files for this, or `Export::Contracts::Row`. I know there's a fair amount of overlap with the `Invoice` equivalents, but do you think it might be helpful to have some in place for regression protection and also to give a developer confidence when making changes to these in the future? appreciate they may end up very similar to the `Invoice` ones, but I don't think that's necessarily a bad thing.

File: spec/lib/tasks/export/contracts_spec.rb [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    let(:output_filename) { '/tmp/contracts_2018-12-25.csv' }
+    let(:args)            { {} }
+    let(:output_lines)    { File.read(output_filename).split("\n") }
+
+    around(:example) do |example|
+      travel_to(Date.new(2018, 12, 25)) { example.run }
+    end
+
+    before do
+      expect(contract.created_at).to be < contract2.created_at
```
Comment (tekin) [inline spec/lib/tasks/export/contracts_spec.rb]: feels a bit weird having an expectation in the `before` block. is this also redundant because of the other expectations that check `output_lines[]`?

File: spec/lib/tasks/export/invoices_spec.rb [hunk 5]
```diff
@@ -1,40 +1,72 @@
...
+    end
 
     before do
-      allow(Export::Invoices::Extract).to receive(:all_relevant).and_return(invoices_to_export)
-      allow(Export::Invoices).to receive(:new).with(
-        invoices_to_export, duck_type(:puts)
-      ).and_return(
-        invoice_exporter
-      )
+      expect(invoice.created_at).to be < invoice2.created_at
```
Comment (tekin) [inline spec/lib/tasks/export/invoices_spec.rb]: same question as above.

File: spec/models/export/anything_spec.rb [hunk 6]
```diff
@@ -0,0 +1,82 @@
+require 'rails_helper'
+
+RSpec.describe Export::Anything do
```
Comment (tekin) [inline spec/models/export/anything_spec.rb]: I'd love to have a chat about the specs for this class. I found it very hard to follow, I think mostly because so much of it is relying on stubs and mocks. it also makes me wonder how much it is coupled to the underlying implementation, and whether that makes them particularly brittle. This may be down to experience and personal preference, but I always feel nervous when I see a spec file were all the actors have been stubbed/mocked. As a hypothetical, how much less confidence would we have in the test suite's coverage of the codebase if this spec file didn't exist at all? just wondering because we have the task specs covering each of the concrete examples.

File: spec/lib/tasks/export/contracts_spec.rb (hunk 4 above)
Comment (rgarner) [reply spec/lib/tasks/export/contracts_spec.rb]: It's a precondition that's saying "this test is order-dependent and if those `let` blocks above aren't respecting that I'm going to tell you".

Given that I'll stick a custom failure message on the end to make that (executably) explicit.

File: app/models/export/contracts/extract.rb (hunk 3 above)
Comment (rgarner) [reply app/models/export/contracts/extract.rb]: I will add one.

File: app/models/export/anything.rb (hunk 2 above)
Comment (rgarner) [reply app/models/export/anything.rb]: Will do.

File: spec/models/export/anything_spec.rb (hunk 6 above)
Comment (rgarner) [reply spec/models/export/anything_spec.rb]: I'm going to either make this go away as a "file deleted, these were design-time specs" commit, or try and replace the stubs with `FactoryBot.create`  a bit, because as you point out, stubbing `:each` is a little bit too far here, and there's a block of four stubs that I have to take some time to parse...

File: app/models/export/contracts/extract.rb (hunk 3 above)
Comment (rgarner) [reply app/models/export/contracts/extract.rb]: I have added an `extract_spec`. Reviewing `invoices/row_spec.rb` it seems that that file is stuffed with design-time specs that helped in mapping when `value_for` didn't exist, but that those tests are not adding much value now. Probably a better thing would be to delete `invoices/row_spec.rb` in favour of adding a small spec for `SubmissionEntryRow`. What do you think?

Comment (tekin) [reply app/models/export/contracts/extract.rb]: sounds good to me.

File: spec/models/export/anything_spec.rb (hunk 6 above)
Comment (rgarner) [reply spec/models/export/anything_spec.rb]: Given that I actually used the spec to update `Export::Anything` to tell us about the class that's missing when there are no records, it's probably useful, so I've de-`double`d it. It comes at the expense of a little factory complexity but I hope it's a bit less of a slog!

File: app/models/export/contracts/extract.rb (hunk 3 above)
Comment (rgarner) [reply app/models/export/contracts/extract.rb]: Done in https://github.com/dxw/DataSubmissionServiceAPI/pull/73/commits/6821b9a6e73e6d4cc98c466cc4f058e4bd10b772
